pkg/reconciler/service/resources/rolloutorchestrator.go

```diff
@@ -0,0 +1,202 @@
+/*
+Copyright 2023 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"fmt"
+	"strconv"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/kmeta"
+	"knative.dev/pkg/ptr"
+	v1 "knative.dev/serving-progressive-rollout/pkg/apis/serving/v1"
+	"knative.dev/serving/pkg/apis/autoscaling"
+	"knative.dev/serving/pkg/apis/serving"
+	servingv1 "knative.dev/serving/pkg/apis/serving/v1"
+)
+
+var (
+	// OverSubRatio is the parameter, that determines how much percentage of the traffic to shift
+	// from the old to the new revision during each stage in the progressive rollout.
+	OverSubRatio = 10
+)
+
+// RevisionRecord is a struct that hosts the name, minScale and maxScale for the revision.
+type RevisionRecord struct {
+	MinScale *int32
+	MaxScale *int32
+	Name     string
+}
+
+// ReadIntAnnotation reads the int value of a specific key in the annotation of the revision.
+func ReadIntAnnotation(revision *servingv1.Revision, key string) (result *int32) {
+	if val, ok := revision.Annotations[key]; ok {
+		i, err := strconv.ParseInt(val, 10, 32)
+		if err == nil {
+			result = ptr.Int32(int32(i))
+		}
+	}
+	return
+}
+
+// ReadIntServiceAnnotation reads the int value of a specific key in the annotation of the service.
+func ReadIntServiceAnnotation(service *servingv1.Service, key string) (result *int32) {
+	if val, ok := service.Spec.Template.ObjectMeta.Annotations[key]; ok {
+		i, err := strconv.ParseInt(val, 10, 32)
+		if err == nil {
+			result = ptr.Int32(int32(i))
+		}
+	}
+	return
+}
+
+// ReadIntRevisionRecord reads the minScale and maxScale in the RevisionRecord.
+func ReadIntRevisionRecord(val RevisionRecord) (min *int32, max *int32) {
+	if val.MinScale != nil {
+		min = ptr.Int32(*val.MinScale)
+	}
+	if val.MaxScale != nil {
+		max = ptr.Int32(*val.MaxScale)
+	}
+	return min, max
+}
+
+func initializeTargetRevisions(ultimateRevisionTarget *[]v1.TargetRevision, traffic *servingv1.TrafficTarget,
```

Comment:
`ultimateRevisionTarget` is not accurate, we are also passing initial revision target

Reply:
suggest renaming to revisionTarget